src/storage/src/healthcheck.rs

```diff
@@ -118,24 +125,181 @@ pub(crate) struct ObjectHealthConfig {
+    /// same worker.
+    SideChannel(String),
+}
+
+impl fmt::Display for StatusNamespace {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use StatusNamespace::*;
+        match self {
+            Kafka => write!(f, "kafka"),
+            Postgres => write!(f, "postgres"),
+            Generator => write!(f, "generator"),
+            TestScript => write!(f, "testscript"),
+            Upsert => write!(f, "upsert"),
+            Decode => write!(f, "decode"),
+            Internal => write!(f, "internal"),
+            StatusNamespace::SideChannel(n) => {
+                write!(f, "{}", n)
+            }
+        }
+    }
+}
+
+struct PerWorkerHealthStatus {
+    pub(crate) errors_by_worker: Vec<BTreeMap<StatusNamespace, HealthStatusUpdate>>,
+}
+
+impl PerWorkerHealthStatus {
+    fn merge_update(
+        &mut self,
+        mut worker: usize,
+        namespace: StatusNamespace,
+        update: HealthStatusUpdate,
+        only_greater: bool,
+    ) {
+        if let StatusNamespace::SideChannel(_) = &namespace {
+            worker = 0;
+        }
+
+        let errors = &mut self.errors_by_worker[worker];
+        match errors.entry(namespace) {
+            Entry::Vacant(v) => {
+                v.insert(update);
+            }
+            Entry::Occupied(mut o) => {
+                if !only_greater || o.get() < &update {
+                    o.insert(update);
+                }
+            }
+        }
+    }
+
+    fn decide_status(&self) -> OverallStatus {
+        let mut output_status = OverallStatus::Starting;
+        let mut namespaced_errors: BTreeMap<String, String> = BTreeMap::new();
+        let mut hints: BTreeSet<String> = BTreeSet::new();
+
+        for status in self.errors_by_worker.iter() {
+            use HealthStatusUpdate::*;
+            for (ns, ns_status) in status.iter() {
+                if let Stalled { error, hint, .. } = ns_status {
+                    let ns = ns.to_string();
+
+                    if Some(error) > namespaced_errors.get(&ns).as_deref() {
+                        namespaced_errors.insert(ns, error.to_string());
+                    }
+
+                    if let Some(hint) = hint {
+                        hints.insert(hint.to_string());
+                    }
+                }
+
+                if !matches!(ns, StatusNamespace::SideChannel(_))
+                    && matches!(ns_status, HealthStatusUpdate::Running)
+                {
+                    output_status = OverallStatus::Running;
+                }
+            }
+        }
+
+        if !namespaced_errors.is_empty() {
+            output_status = OverallStatus::Stalled {
+                // Concatenate the namespaced errors together with commas, for an individual
+                // error message the user can look at.
+                error: namespaced_errors
+                    .iter()
+                    .map(|(ns, err)| format!("{}: {}", ns, err))
+                    .join(", "),
```

Comment:
Yeah, on reflection, I still think my preference is: decide some total order over the namespaces, include the "earliest" error here, and remove it from the map so it's not in both the details section and the error section.

To mehat seems as good or better from the user's perspective, and it avoids a bunch of downstream work...

Reply:
I implemented this!